Switch formatting from black to ruff-format

akx opened the pull request:
## Breaking change (for developers only)

* ~There is no `"python.formatting.provider": "ruff"` configuration for VSCode users.~

## Proposed change

This proposes to switch the formatting for the project from `black` to `ruff-format` ([released in October](https://astral.sh/blog/the-ruff-formatter)).

The Ruff folks are tracking home-assistant's code style c.f. that of `ruff-format`'s, and [it's roughly 99.9% compatible](https://github.com/astral-sh/ruff/actions/runs/6661598076#summary-18104703550) at this point.

You can find the proposed reformatting diff at 2dd0a74b38ecb454762850b1e0c4aba9da2d0986 in this gist: https://gist.github.com/akx/89e5d6ee7bfeda70a2466c1d591d2922 (updated).
To my eye, the changes seem pretty superficial.

As before with Ruff, the big win is performance. There'll be no need to faff around with a Black cache anymore in CI (and the entire partial/full checks could be unified to "just always check everything").

This `hyperfine`ing is on a pretty beefy 12-core M2 Max Macbook, FWIW. Caches are disabled to level the playground.

```
$ env BLACK_CACHE_DIR=/dev/null hyperfine -N -i 'black --check .' 'ruff format --no-cache --check .'
Benchmark 1: black --check .
  Time (mean ± σ):     36.258 s ±  6.661 s    [User: 242.110 s, System: 4.790 s]
  Range (min … max):   27.607 s … 45.987 s    10 runs

Benchmark 2: ruff format --no-cache --check .
  Time (mean ± σ):     418.4 ms ±   9.4 ms    [User: 3603.6 ms, System: 311.9 ms]
  Range (min … max):   407.5 ms … 436.3 ms    10 runs

  Warning: Ignoring non-zero exit code.

Summary
  ruff format --no-cache --check . ran
   86.66 ± 16.04 times faster than black --check .
$
```

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Checklist

- [x] The code change is tested and works locally.
  - Aside from a warning:
    > The following rules may cause conflicts when used with the formatter: `ISC001`.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
  - For obvious reasons, nope... 😁  
- [ ] Tests have been added to verify that the new code works.

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

akx committed:
Switch formatting from black to ruff-format

akx commented:
Sorry for the ping, but WDYT @frenck? Does this have a fighting chance at this point, or should I let it brew for a while? 😄

MichaReiser commented on pyproject.toml:
```diff
@@ -703,3 +700,6 @@ split-on-trailing-comma = false
 max-complexity = 25
+
+[tool.ruff.format]
+exclude = ["/generated/"]
```
It may be necessary to change this to `/generated/**` See https://github.com/astral-sh/ruff/issues/8267

akx replied:
In fact, I think this is unnecessary altogether, since the generated files do get filtered through `black` at generation time. I suspect it's just here because there hasn't been a reason to spend cycles on Black re-checking them... whereas with Ruff, well...

frenck commented:
I've been deploying this (and actively using it in vscode for Core development as well). I am personally supportive. But I will check with some other core team members on how they feel about this.

Additionally, if this is merged, this should be done a few days before the next beta, preferably after the last patch release has been shipped for the current cycle. This is to prevent merge conflicts and differences in CI runtime between those.

frenck commented:
@akx We have agreed on moving to Ruff for formatting.

I suggest waiting for the release of 2023.11.3 before moving `dev` towards it (the release of 2023.11.3 is planned for today 21 November)

akx commented:
@frenck Wow, excellent timing, I just rebased this 😂

frenck commented:
@akx 2023.11.3 has been cut and shipped.

akx committed:
Apply ruff-format

akx commented:
(There I thought I was being clever attributing the actual formatting change commit to Ruff... 😄)

akx commented:
@frenck Yay! Rebased on current `dev`, added the actual reformatting commit. Feel free to un-draft & merge & so on 😊

akx commented:
@frenck Ah, yeah, I'll adjust ignore comments back where mypy expects them to be...

frenck commented on homeassistant/components/homekit/type_fans.py:
```diff
@@ -127,9 +127,8 @@ def __init__(self, *args: Any) -> None:
-                    setter_callback=lambda value, preset_mode=preset_mode: self.set_preset_mode(
-                        value, preset_mode
-                    ),
+                    setter_callback=lambda value,
+                    preset_mode=preset_mode: self.set_preset_mode(value, preset_mode),
```
We need to wrap the lambda in parenthesis, as multiline lambdas is not something we generally want as they are hard to read.

frenck commented on homeassistant/components/hunterdouglas_powerview/select.py:
```diff
@@ -116,5 +116,7 @@ def current_option(self) -> str | None:
-        await self._shade.refresh()  # force update data to ensure new info is in coordinator
+        await (
+            self._shade.refresh()
+        )  # force update data to ensure new info is in coordinator
```
I think the comment should be moved two lines up

frenck commented on homeassistant/components/matter/event.py:
```diff
@@ -104,9 +104,7 @@ def _update_from_device(self) -> None:
-    def _on_matter_node_event(
-        self, event: EventType, data: MatterNodeEvent
-    ) -> None:  # noqa: F821
+    def _on_matter_node_event(self, event: EventType, data: MatterNodeEvent) -> None:  # noqa: F821
```
This seems like an odd change. Instead of reformatting, the noqa command should have been moved.

frenck commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -470,9 +470,7 @@ def handle_climate_attribute_received(
-    def prepare_subscribe_topics(
-        self, topics: dict[str, dict[str, Any]]
-    ) -> None:  # noqa: C901
+    def prepare_subscribe_topics(self, topics: dict[str, dict[str, Any]]) -> None:  # noqa: C901
```
This with the comment causes a longer line. We should have moved the noqa comment instead of having this formatted.

frenck commented on homeassistant/components/private_ble_device/sensor.py:
```diff
@@ -83,7 +83,8 @@ class PrivateDeviceSensorEntityDescription(
         entity_registry_enabled_default=False,
         entity_category=EntityCategory.DIAGNOSTIC,
-        value_fn=lambda hass, service_info: bluetooth.async_get_learned_advertising_interval(
+        value_fn=lambda hass,
```
This casuses a multiline lambda, which should add some parenthesis, instead of accepting this formatting

akx committed:
Adjust mypy comments back to where they belong

frenck commented on homeassistant/components/recorder/auto_repairs/schema.py:
```diff
@@ -101,9 +101,7 @@ def _validate_table_schema_has_correct_collation(
-            or dialect_kwargs.get(
-                "mariadb_collate"
-            )  # pylint: disable-next=protected-access
+            or dialect_kwargs.get("mariadb_collate")  # pylint: disable-next=protected-access
```
`disable-next` seems a bit oddly placed now, this comment should move a bit.

akx replied:
I think this is correct – the next line has the `_fetch_setting` private use (just like it did before).

frenck replied:
Right, but we can place it on its own line. Right now it implies a bit the comment is about the line it is on :)

frenck commented on homeassistant/components/stream/recorder.py:
```diff
@@ -78,7 +78,13 @@ async def async_record(self) -> None:
         def write_segment(segment: Segment) -> None:
             """Write a segment to output."""
-            nonlocal output, output_v, output_a, last_stream_id, running_duration, last_sequence
+            nonlocal \
+                output, \
+                output_v, \
+                output_a, \
+                last_stream_id, \
+                running_duration, \
+                last_sequence
```
Ouch, for me the use of `\` is a breaking deal here 😅

akx replied:
Would that `nonlocal (..., ..., ...)` was valid syntax 😄 

That said, maybe the use of `nonlocal` here is something to look at in general; it smells like these should be instance variables :)

frenck commented on homeassistant/util/location.py:
```diff
@@ -138,11 +138,7 @@ def vincenty(
-                * cos2SigmaM
-                * (-3 + 4 * sinSigma**2)
-                * (-3 + 4 * cos2SigmaM**2)
+                - B / 6 * cos2SigmaM * (-3 + 4 * sinSigma**2) * (-3 + 4 * cos2SigmaM**2)
```
This harms readability, lets disable formatting around this part.

akx replied:
That function doesn't adhere much to the surrounding code style either, but sure thing, adding some strategic `# fmt: ...`s :)

akx committed:
Adjust based on review

frenck commented:
We should update the devcontainer and vscode ext to include the ruff extension and set it as the default formatter for this project.

akx committed:
Adjust test patches with trailing commas for readability and to minimize formatting patch

akx committed:
tests/zwave_js: use f-strings for neater formatting

akx commented:
@frenck Good to review again, I think... 😄 

* I don't know much at all about VSCode/devcontainers (IntelliJ tools all the way here 😁), so I'm not sure where to go with that to make sure everything goes smoothly. Could be done in a subsequent PR?
* I fixed up the remaining review comments and gave the files another manual look-over to fix some remaining ugly lambdas and so on (and changed some `.format()`ting to f-strings in the zwave tests so they get formatted more nicely).
* With the assumption this will be squash-merged, there's also the last commit in this PR where I adjusted a bunch of trailing commas in the `patch()` litanies in the tests, to minimize the total size of the squashed merge ✨

frenck commented:
> I don't know much at all about VSCode/devcontainers (IntelliJ tools all the way here 😁), so I'm not sure where to go with that to make sure everything goes smoothly. Could be done in a subsequent PR?

I'll take a look

frenck commented:
@akx I can't push to this branch apparently, but this is the diff:

```diff
diff --git a/.devcontainer/devcontainer.json b/.devcontainer/devcontainer.json
index 83ee0a2e422..44a81718e10 100644
--- a/.devcontainer/devcontainer.json
+++ b/.devcontainer/devcontainer.json
@@ -10,7 +10,7 @@
   "customizations": {
     "vscode": {
       "extensions": [
-        "ms-python.black-formatter",
+        "charliermarsh.ruff",
         "ms-python.pylint",
         "ms-python.vscode-pylance",
         "visualstudioexptteam.vscodeintellicode",
@@ -39,7 +39,10 @@
           "!include_dir_list scalar",
           "!include_dir_merge_list scalar",
           "!include_dir_merge_named scalar"
-        ]
+        ],
+        "[python]": {
+          "editor.defaultFormatter": "charliermarsh.ruff"
+        }
       }
     }
   }
diff --git a/.vscode/extensions.json b/.vscode/extensions.json
index 951134133e5..8a5d7d486b7 100644
--- a/.vscode/extensions.json
+++ b/.vscode/extensions.json
@@ -1,3 +1,7 @@
 {
-  "recommendations": ["esbenp.prettier-vscode", "ms-python.python"]
+  "recommendations": [
+    "charliermarsh.ruff",
+    "esbenp.prettier-vscode",
+    "ms-python.python"
+  ]
 }
```

frenck committed:
Adjust devcontainer config

frenck reviewed (approved):
Looks good to go to me! 👍

frenck reviewed (approved):
Good to merge once the CI passes.

bdraco reviewed (approved):
Nice
